### PR TITLE
server: preemption notices, asynchronous parks and exact concurrency (pin branch on b10869)

### PR DESCRIPTION
**Pull request by danielhanchen**
The tree of #197 (its branch merged with the upstream release tag b10871) composed with the pinned commit of ggml-org#25731 (inkling, 946fc11d1a), so the nightly's pin merge has one merge base after inkling.

History of this branch:

- afd3248b8: #197's tree as a single commit on b10869. #197's own history contains fork master's history, which shares ancestors with the pinned branches, and the composed merge (tag + the pins in scripts/unsloth/pr-set.json, in order) then had three merge bases and spurious conflicts in files #197 never touched.
- 2834f07a9: merged with the inkling pin. Inkling adds `f32_pedantic` guards inside the mul_mat dispatch #197 restructured into `ggml_cuda_mul_mat_path`, and adds neighbours in llama-kv-cache.h and test-backend-ops.cpp at the same insertion points. The guard now lives in `ggml_cuda_mul_mat_path`, so the switch, the invariant width and the column split pick the same implementation; `ggml_cuda_mul_mat_id_needs_sync` keeps both `!f32_pedantic` and #197's `ntokens`; the flash attention precision call uses `ggml_prec_set_acc`, which inkling introduces in place of the deprecated setter. CUDA build warning-free; test-exact-geometry, test-exact-buft, test-exact-pages pass; test-backend-ops MUL_MAT 1312/1312 and MUL_MAT_ID 930/930 on CUDA.
- 62cf50230: drops the 40 fork-owned files (unsloth-*.yml workflows, the prebuilt-alert action, scripts/unsloth) that #197's branch point carried; the nightly composes those from master and the other pins carry none.

Reproduced the preflight's resolve loop in a fresh clone of upstream on b10871 (temp replay of unsloth-pin-preflight.yml: fetch each pin, `merge --no-ff` with diff3, additive_merge.py on conflict): the thirteen current pins merge as the preflight reports, then this commit merges with two pure add/add hunks against earlier pins (the proc-address table entry next to #24423's, the exact-concurrency refusals next to the kpool-dirty blocks in llama-memory-hybrid.cpp), which additive_merge.py resolves. merge_checks clean (23 python and 189 c++ files); pin_contract: all 14 pins intact, this one 25 files with new symbols and 6559 lines accounted for.

The base branch upstream/b10871-inkling-25731 is the inkling pin pushed as a branch, so the diff here is #197's change set on top of inkling. Review happens on #197; this branch is only what gets pinned. Because the merge base is inkling's pinned commit, this pin has to be recomposed whenever ggml-org#25731 is repinned or leaves the set.
- 1cfb42bf5, 17a0f3b96: cherry-picks of #197's 50aa4be21 (a test that a refused exact-mode cross-sequence copy leaves both sequences as they were) and 4700c7f7c (the allocation granularity comment condensed). #197's 15df6351b (ggml_prec_set_acc, which b10871 itself deprecates the flash attention setter in favour of) is already in 2834f07a9.
- 1d5548978: cherry-pick of #197's fb5f029f7. `--preempt-ram` now defaults to 0, so a server launched without the flag behaves exactly as upstream, unified cache included: no planner, no pinned probe, no async copier, no startup notice, and the cache-full path fails every slot as upstream does (verified on GPU with four concurrent completions past a 512-cell pool: four context errors and zero preemption lines without the flag; four completions and 38 park and resume lines with `--preempt-ram 1024`). Exact concurrency stays an environment opt-in. Replay is clean on b10871 and on b10884, the newest aged tag at the time.
- 0b9caab11, 15bcc6c7a: cherry-picks of #197's 04bcb0aa8 and 2b27c22f4. The paged flash attention path now reads its page table from `src[5]` only for `GGML_OP_FLASH_ATTN_EXT`, since inkling's banded op keeps its relative bias in the same slot (reverting the gate in the composed tree aborts on the banded MMA path; `FLASH_ATTN_EXT_BANDED` 13/13 with it). The sequence-plane normalisation follows `GGML_CUDA_BATCH_INVARIANT` as its comment promises, not only the exact-mode variable. Replay clean on b10871 and b10884.

### CODE REVIEW

File: .github/workflows/unsloth-prebuilt.yml [hunk 1]
```diff
@@ -0,0 +1,1317 @@
...
+              echo "jobs API call failed (${API_FAILS}/10); retrying in ${POLL}s"
+              sleep "$POLL"
+              continue
+            fi
+            API_FAILS=0
+
+            # Everything in the run except this job and the alert job that
+            # reports on it. `resolve` stays in the set; it is already a
+            # `needs:`, so it is a free consistency check.
+            SIBS="$(printf '%s\n' "$JOBS" | awk -F'\t' -v self="$SELF" -v alert="$ALERT" 'NF && $1 != self && $1 != alert')"
```
[the remark this answers is not in the document]
Comment (danielhanchen) [reply .github/workflows/unsloth-prebuilt.yml]: These workflow files were stale copies of the fork's own nightly workflows carried in from the branch point, not part of the pin. 62cf50230 drops all 40 fork-owned files (unsloth-*.yml, the prebuilt-alert action, scripts/unsloth) from the pin tree; the nightly composes those from master. The three items on those files are therefore moot here; if the matrix waiter or the ARM label are wrong on master they belong in a master PR.

File: .github/workflows/unsloth-prebuilt.yml [hunk 2]
```diff
@@ -0,0 +1,1317 @@
...
+            echo "### Artifact storage reclaimed (unpublished run)"
+            echo ""
+            echo "| metric | value |"
+            echo "| --- | --- |"
+            echo "| artifacts deleted | $deleted |"
+            echo "| storage freed | $(( freed / 1048576 )) MiB |"
+            echo "| delete failures | $failed |"
+          } >> "$GITHUB_STEP_SUMMARY"
+
+      # ── Keep the ccache budget inside the repo limit ──
```
[the remark this answers is not in the document]
Comment (danielhanchen) [reply .github/workflows/unsloth-prebuilt.yml]: Moot after 62cf50230: this file was a stale copy of the fork's own workflow carried in from the branch point and is no longer in the pin tree. See the reply on the matrix-waiter thread.

File: .github/workflows/unsloth-prebuilt-cpu.yml [hunk 3]
```diff
@@ -0,0 +1,376 @@
...
+jobs:
+  build-linux:
+    name: linux/${{ matrix.arch }}
+    runs-on: ${{ matrix.runner }}
+    strategy:
+      fail-fast: false
+      matrix:
+        include:
+          - { arch: x64,   runner: ubuntu-22.04 }
+          - { arch: arm64, runner: ubuntu-22.04-arm }
```
[the remark this answers is not in the document]
Comment (danielhanchen) [reply .github/workflows/unsloth-prebuilt-cpu.yml]: Moot after 62cf50230: this file was a stale copy of the fork's own workflow carried in from the branch point and is no longer in the pin tree. See the reply on the matrix-waiter thread.

File: src/llama-kv-cache.cpp [hunk 4]
```diff
@@ -454,6 +675,14 @@ void llama_kv_cache::seq_cp(llama_seq_id seq_id_src, llama_seq_id seq_id_dst, ll
         return;
     }
 
+    // [TAG_EXACT_CONCURRENCY] a page belongs to one sequence, so refuse a copy that would share cells rather than abort. After the shared-cells return, so a draft cache is unaffected.
+    if (exact_pages && seq_id_src != seq_id_dst) {
+        LLAMA_LOG_ERROR("%s: exact concurrency does not support copying cells between "
+                        "sequences (%d -> %d); ignoring the copy\n",
+                        __func__, seq_id_src, seq_id_dst);
+        return;
```
[the remark this answers is not in the document]
Comment (danielhanchen) [reply src/llama-kv-cache.cpp:683]: The copy is not discarded silently: llama_kv_cache::seq_cp logs LLAMA_LOG_ERROR("exact concurrency does not support copying cells between sequences (%d -> %d); ignoring the copy") before returning, and llama_memory_hybrid::seq_cp does the same. llama_memory_seq_cp returns void, so an error log is the only signal the API allows, and aborting is what that path deliberately avoids. Rejecting before the caller enters the flow is already done for the only caller this repo ships: seq_cp is reached solely from server_slot::copy_state_to, whose only caller is the parent/child family loop, and a family only exists for n > 1, which exact mode turns into a 400 before the task is created. The upstream examples that copy a shared prefix predate the mode and I would rather not rewire four of them for an opt-in environment variable. What was missing is a regression test, so test-exact-pages now asserts that a refused copy leaves the destination empty and the source unchanged (unslothai/llama.cpp#197 50aa4be21, carried here as 1cfb42bf5), confirmed on a model that can run the mode.

File: include/llama.h [hunk 5]
```diff
@@ -804,6 +804,27 @@ extern "C" {
     // Check if the memory supports shifting
     LLAMA_API bool llama_memory_can_shift(llama_memory_t mem);
 
+    // [TAG_EXACT_CONCURRENCY] cells the memory allocates in one indivisible unit: 1 ordinarily,
+    // larger where a mode places cells in blocks, and then n contiguous tokens occupy
+    // round_up(n, granularity) cells. A caller deciding whether the pool has room must round the
```
[the remark this answers is not in the document]
Comment (danielhanchen) [reply include/llama.h]: Condensed to two lines in #197 (4700c7f7c, carried here as 17a0f3b96), keeping the granularity definition, the round_up rule for contiguous tokens and the caveat that a sequence left with holes still holds every block one of its live cells sits in. The worked example is dropped; there is no docs page for this feature to move it to.

File: tests/CMakeLists.txt [hunk 6]
```diff
@@ -333,6 +333,15 @@ llama_build_and_test(test-backend-sampler.cpp   LABEL "model")
 llama_build_and_test(test-state-restore-fragmented.cpp LABEL "model" ARGS -m "${MODEL_DEST}")
 set_tests_properties(test-state-restore-fragmented PROPERTIES FIXTURES_REQUIRED test-download-model)
 
+# Guards on the asynchronous per-sequence state transfer
+# Skips itself on a backend that cannot copy asynchronously
+llama_build_and_test(test-state-seq-copy.cpp LABEL "model" ARGS -m "${MODEL_DEST}")
+set_tests_properties(test-state-seq-copy PROPERTIES FIXTURES_REQUIRED test-download-model)
```
[the remark this answers is not in the document]
Comment (danielhanchen) [reply tests/CMakeLists.txt:341]: As the fork maintainer I approve these five standalone tests. test-exact-geometry is pure batch-shape arithmetic that links only common and needs no model or backend, so folding it into test-exact-pages would lose it on model-less and CPU-only CI, and test-backend-ops has no llama context at all; test-exact-buft needs ggml-backend plus the internal src/llama-impl.h and a device but no model, a third combination outside test-backend-ops; test-exact-pages needs a real context and a model and sits behind the download fixture under LABEL model; test-server-tokens links server-context and mtmd with the tools/server include path under LLAMA_BUILD_TOOLS, which no file in tests/ otherwise does; test-state-seq-copy drives llama_state_seq_copy_* on a real context, which test-state-restore-fragmented does not cover.

File: src/llama-context.cpp [hunk 7]
```diff
@@ -3062,6 +3335,273 @@ size_t llama_context::state_set_data(const uint8_t * src, size_t size) {
...
+            return nullptr;
+        }
+
+        host_buf.reset(buf);
+
+        data     = base;
+        size     = size_new;
+        capacity = size_new;
+        // a host buffer type may quietly hand back ordinary memory when pinning is off, so believe the buffer that came back rather than the type
+        pinned   = can_pin && ggml_backend_buffer_get_type(buf) == host_buft;
```
[the remark this answers is not in the document]
Comment (danielhanchen) [reply src/llama-context.cpp:3499]: This is upstream ggml code the PR does not touch, and the flag is not needed. ggml_cuda_host_malloc calls cudaMallocHost, and the CUDA Runtime API's Unified Addressing section states that all host memory allocated through cudaMallocHost() and cudaHostAlloc() is always directly accessible from all devices that support unified addressing, regardless of whether cudaHostAllocPortable or cudaHostAllocMapped was specified, and that unified addressing is automatically enabled in 64-bit processes (https://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__UNIFIED.html); the Programming Guide's Portable Memory section says the same, that the default benefits cover the allocating device and all devices sharing the unified address space. Every configuration the CUDA backend supports is a 64-bit process, so the buffer is portable everywhere the server runs and pinned = true is accurate. It is also the existing pattern: llama_context already takes the host buffer of the first device for intermediate state across a multi-device split, and the model loader picks the first device offering a host buffer type, so a real gap here would be an upstream defect in split mode generally.

File: ggml/src/ggml-cuda/fattn.cu [hunk 8]
```diff
@@ -738,6 +743,52 @@ size_t ggml_cuda_flash_attn_ext_get_alloc_size(int device, const ggml_tensor * d
 
 void ggml_cuda_flash_attn_ext(ggml_backend_cuda_context & ctx, ggml_tensor * dst) {
     ggml_cuda_set_device(ctx.device);
+
+    if (dst->src[5]) {
+        GGML_ASSERT(dst->src[0]->ne[0] == 256 && dst->src[2]->ne[0] == 256);
```
[the remark this answers is not in the document]
Comment (danielhanchen) [reply ggml/src/ggml-cuda/fattn.cu]: Confirmed, and it aborts rather than degrades: on a B200 the banded op's MMA path delegates to ggml_cuda_flash_attn_ext, which read src[5] unconditionally and asserted the paged shape. Reverting only the gate in the composed tree reproduces it exactly (fattn.cu:748 GGML_ASSERT on the 256-wide heads, exit 134). Fixed in unslothai/llama.cpp#197 04bcb0aa8, carried here as 0b9caab11: the paged path is taken from one helper, ggml_cuda_fattn_pages(dst), which returns src[5] only when dst->op == GGML_OP_FLASH_ATTN_EXT, and every downstream site in launch_fattn uses it too (the KV_max scan, the one-block-per-tile split, and the indices kernel argument, which would have handed the F32 bias to the kernel as a page table). This mirrors what the banded code already does for its own slot, so the two ops partition src[5] by op and the bias still travels through its own rel_f parameter. FLASH_ATTN_EXT_BANDED goes from that abort to 13/13, with FLASH_ATTN_EXT at 2962/2962 and MUL_MAT at 1312/1312 alongside it.

File: ggml/src/ggml-cuda/ggml-cuda.cu [hunk 9]
```diff
@@ -1846,63 +1855,285 @@ static bool ggml_cuda_should_fuse_mul_mat_vec_q(const ggml_tensor * tensor) {
...
+    return 1;
+}
+
+static bool ggml_cuda_mul_mat_split_columns(
+        ggml_backend_cuda_context & ctx, int cc, int warp_size,
+        const ggml_tensor * src0, const ggml_tensor * src1, ggml_tensor * dst) {
+    // recurrent output projections broadcast one weight matrix over sequence planes, so normalize each plane before applying the column policy
+    if (ggml_cuda_exact_concurrency() && src0->ne[2] == 1 && src0->ne[3] == 1 &&
+            (dst->ne[2] > 1 || dst->ne[3] > 1) &&
+            src1->ne[2] == dst->ne[2] && src1->ne[3] == dst->ne[3]) {
```
[the remark this answers is not in the document]
Comment (danielhanchen) [reply ggml/src/ggml-cuda/ggml-cuda.cu]: Agreed, and the comment is honoured rather than corrected. The documented contract is stated for GGML_CUDA_BATCH_INVARIANT (1 computes every destination column on its own, exactly as a batch of one would), and with the guard on the exact-mode variable a broadcast matmul carrying sequence planes fell through to the generic splitter, which rejects it on ne[2]/ne[3] and leaves it batched. Fixed in unslothai/llama.cpp#197 2b27c22f4, carried here as 15bcc6c7a: the guard reads ggml_cuda_batch_invariant(). This is a narrowing removed rather than a new path: the function is only reachable under ggml_cuda_batch_invariant() in the first place, the decomposition is already guarded on the broadcast shape and matching plane counts, and each plane then goes through the same column policy a solo request would get. MUL_MAT passes 1312/1312 plain and with GGML_CUDA_BATCH_INVARIANT=1 and =2, along with test-exact-geometry.

File: common/common.cpp [hunk 10]
```diff
@@ -1433,6 +1451,148 @@ std::vector<llama_adapter_lora_ptr> & common_init_result::lora() {
...
+}
+
+bool common_exact_concurrency_init(const common_params & params) {
+    if (!common_exact_concurrency()) {
+        return true;
+    }
+
+    // DFlash drafting turns causal attention off on its draft context, which the paged attention needs; say so instead of asserting in the graph. DSpark is the same.
+    for (const auto type : params.speculative.types) {
+        if (type == COMMON_SPECULATIVE_TYPE_DRAFT_DFLASH || type == COMMON_SPECULATIVE_TYPE_DRAFT_DSPARK) {
```
[the remark this answers is not in the document]
Comment (danielhanchen) [reply common/common.cpp:1512]: A shared-memory draft never claims a cell: apply_ubatch returns early when other is set and it is the only caller of exact_pages_claim, so every draft step sees the same free cell at n_past. The combination is also unreachable: ctx_other exists only for GEMMA4_ASSISTANT, EAGLE3 and DFLASH, the last two cannot produce an MTP context, and the Gemma 4 assistant is SWA, which exact mode refuses at cache construction.

**Comment by danielhanchen** [comment]
@codex review

**Comment by danielhanchen** [comment]
@codex security review

**Comment by danielhanchen** [comment]
@codex review

**Comment by danielhanchen** [comment]
@codex review

**Comment by danielhanchen** [comment]
@codex security review

File: tools/server/server-context.cpp [hunk 11]
```diff
@@ -339,6 +412,380 @@ struct server_slot {
...
+            preempt_state_tgt.resize(size_tgt);
+            preempt_state_dft.resize(size_dft);
+        } catch (const std::bad_alloc & e) {
+            SLT_ERR(*this, "failed to allocate %.3f MiB for the preemption state: %s\n",
+                    (size_tgt + size_dft) / (1024.0 * 1024.0), e.what());
+            preempt_state_free();
+            return false;
+        }
+
+        if (llama_state_seq_get_data_ext(ctx_tgt, preempt_state_tgt.data(), size_tgt, id, LLAMA_STATE_SEQ_FLAGS_NONE) != size_tgt) {
```
[the remark this answers is not in the document]
Comment (danielhanchen) [reply tools/server/server-context.cpp:690]: Both synchronous calls already synchronize inside llama.cpp before any KV byte moves: llama_state_seq_get_data_ext and llama_state_seq_set_data_ext call ctx->synchronize() first (llama-context.cpp:4991 and :4996), which is ggml_backend_sched_synchronize over every backend, and on CUDA that is cudaStreamSynchronize on the main stream that the forked graph streams are joined back onto within graph_compute. The skipped synchronize after an output-free batch is recovered there, so a second llama_synchronize before the save or restore would be a duplicate round trip.

**Comment by danielhanchen** [comment]
@codex review

**Comment by danielhanchen** [comment]
@codex review

**Comment by danielhanchen** [comment]
@codex review
